### ISSUE TITLE
Problem with certificates on using Azure-sdk-c on debian-10 and fedora-33

### ISSUE THREAD

**Issue by BinduRao2018** [Azure/azure-iot-sdk-c#2021]
Based on the inputs/discussion in #1900, I have created certificates, register a device, provision and send messages using MQTT on Ubuntu - 20.04. However, the same configurations (device-id, group enrollment license) doesn't work when used on Fedora-33 & Debian-10 though provisioning is successful.

Here is the error log FYI:
Error: Time:Tue Jul 13 15:36:08 2021 File:/cloned/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:mqtt_error_callback Line:139 MQTT communication error
Error: Time:Tue Jul 13 15:36:08 2021 File:/cloned/azure-iot-sdk-c/provisioning_client/src/prov_device_ll_client.c Func:on_transport_registration_data Line:771 Failure retrieving data from the provisioning service
Error: Time:Tue Jul 13 15:36:08 2021 File:/cloned/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:sendPacketItem Line:395 Failure sending control packet data
Error: Time:Tue Jul 13 15:36:08 2021 File:/cloned/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_disconnect Line:1300 Error: mqtt_client_disconnect send failed

Please note, we build azure-sdk-c on the respective systems.

Is there any limitation on the way the certificates are created/used on these systems?

**Comment by BinduRao2018** [comment]
Can I have any update on this, please?

**Comment by BinduRao2018** [comment]
Is anyone using azure-sdk-c on debian-10 or any RPM based systems? Is there any difference in the way certificates are managed on these systems? Could anybody provide inputs/clarify on the above mentioned problem?

**Comment by danewalton** [comment]
Interesting are you seeing different behavior with the same code on different systems? The logs there make it seem like the device is trying to disconnect and can't send the disconnect packet to the provisioning service. Do you have any Wireshark logs you can share for the scenario?

**Comment by BinduRao2018** [comment]
@danewalton, here is the Wireshark log that was captured on Debian-10

[azure-sdk-c-debian10-ws.csv](https://github.com/Azure/azure-iot-sdk-c/files/7122845/azure-sdk-c-debian10-ws.csv)

**Comment by danewalton** [comment]
From what I can tell in the logs your device is sending a `RST` to the server after the Change Cipher Spec step from the server. This could mean your system doesn't have the necessary cipher suites.

You can see here for the supported ones we have:
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-tls-support

**Comment by BinduRao2018** [comment]
I had a look at the available ciphers on both systems (Ubuntu-20.04 & Debian-10). The cipher suites are exactly the same. As mentioned, device provisioning works fine on Ubuntu-20.04, but not on Debian-10

**Comment by hihigupt** [comment]
@BinduRao2018 Can you please share the Wireshark trace as pcapng format ? csv export has very little information.

**Comment by BinduRao2018** [comment]
@hihigupt, please find the logs in the attached zip

[azure-sdk-c-debian10-ws.zip](https://github.com/Azure/azure-iot-sdk-c/files/7139793/azure-sdk-c-debian10-ws.zip)

**Comment by danewalton** [comment]
Are you setting the trusted certs as an option or relying on a system certificate store here? AKA here in samples:
https://github.com/Azure/azure-iot-sdk-c/blob/54ef686d6faa3443ccebf633038789cf0f05df7e/provisioning_client/samples/prov_dev_client_ll_sample/prov_dev_client_ll_sample.c#L221

**Comment by BinduRao2018** [comment]
No, I haven't set any option and rely on the certificate store.

**Comment by danewalton** [comment]
Gotcha okay would you mind setting the option and running it again to see if it works? If it works with the cert option we can assume it is likely the cert store and can then go from there.

**Comment by BinduRao2018** [comment]
I tried to enable the 'OPTION_TRUSTED_CERT' option as suggested, but see the same error on Debian-10, however works fine on Ubuntu-20.04.
I have used certificates from certs.c available in the repository. Is that correct?

**Comment by danewalton** [comment]
Yup that's correct. Do you have the full logs for when this runs? Also can you give details on how the sample is configured (custom hsm vs riot etc).

**Comment by BinduRao2018** [comment]
The log that I have shared earlier is the only information available. 
i.e.
=========================================================================================
Error: Time:Mon Sep 13 18:02:17 2021 File:/cloned/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:mqtt_error_callback Line:139 MQTT communication error
Error: Time:Mon Sep 13 18:02:17 2021 File:/cloned/azure-iot-sdk-c/provisioning_client/src/prov_device_ll_client.c Func:on_transport_registration_data Line:771 Failure retrieving data from the provisioning service
Error: Time:Mon Sep 13 18:02:17 2021 File:/cloned/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:sendPacketItem Line:395 Failure sending control packet data
Error: Time:Mon Sep 13 18:02:17 2021 File:/cloned/azure-iot-sdk-c/umqtt/src/mqtt_client.c Func:mqtt_client_disconnect Line:1300 Error: mqtt_client_disconnect send failed
=========================================================================================
Is there any way to get more information?

Please note,  call to 'Prov_Device_LL_Register_Device()', return the reg_status a - 'CONNECTED' in the callback. However, registration is not successful.

I have defined the v-table for the x509 hsm calls and pass the full-chain certificate to the hsm_info_global->certificate.

**Comment by danewalton** [comment]
Interesting.....I ask because for example [in this issue](https://github.com/Azure/azure-iot-sdk-c/issues/1900#issuecomment-800476118) there are additional logs such as 

>Provisioning Status: PROV_DEVICE_REG_STATUS_CONNECTED
Error: Time:Tue Mar 16 17:42:12 2021 File:/home/br/workbench/Azuresphere/azure-iot-sdk-c/provisioning_client/src/prov_device_ll_client.c Func:prov_transport_process_json_reply Line:670 Unsuccessful json encountered: {"errorCode":401002,"trackingId":"fa14c8d0-29f8-49ab-aa0e-31892282bfd1","message":"CA certificate not found.","timestampUtc":"2021-03-16T17:42:12.9197645Z"}
Error: Time:Tue Mar 16 17:42:12 2021 File:/home/br/workbench/Azuresphere/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:prov_transport_common_mqtt_dowork Line:1014 Unable to process registration reply.
Error: Time:Tue Mar 16 17:42:12 2021 File:/home/br/workbench/Azuresphere/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:mqtt_error_callback Line:139 MQTT communication error
Error: Time:Tue Mar 16 17:42:12 2021 File:/home/br/workbench/Azuresphere/azure-iot-sdk-c/provisioning_client/src/prov_device_ll_client.c Func:on_transport_registration_data Line:771 Failure retrieving data from the provisioning service

which give some extra clues into what might be going wrong.

From the original wireshark, the TLS handshake looks to go through and there is application data which looks to be sent.

![image](https://user-images.githubusercontent.com/36710865/134063035-4830a480-daf7-4ddc-ab02-e31b4ef1829a.png)


Can you describe in detail what "full chain" you have? Which certs are registered to the provisioning service and which are stored on the device? If you have the full source of the hsm you created that could also be helpful.

**Comment by BinduRao2018** [comment]
I use a full-chain as a certificate - i.e Root CA, intermediate and leaf certificate. The keys & certificates are generated as described in https://github.com/Azure/azure-iot-sdk-c/blob/master/tools/CACertificates/CACertificateOverview.md#managing-test-ca-certificates-for-samples-and-tutorials.

A private key and full-chain certificates are passed when hsm_get_key and hsm_get_certificate functions are called.  hsm implementation is similar to custom_hsm_example.c available in the repository. get_key() and get_certificate() return the same certificate (full-chain) and key for both calls invoked for the device and provision service.

**Comment by danewalton** [comment]
Okay I was able to get a Debian machine up and running and ran the provisioning ll sample with a custom HSM and cert. It looks like everything is working properly so my hunch is that you have misconfigured your certificates.

![image](https://user-images.githubusercontent.com/36710865/134596837-422b2863-6649-47cf-b998-07dacfcaddbe.png)

Can you please try using our embedded directions below to generate a self signed cert and use that? I know it's not the exact scenario I just want to hone in on the culprit.

https://github.com/Azure/azure-sdk-for-c/tree/main/sdk/samples/iot#linux-certificate-creation
